src/graphql-types/ReasonApolloQuery.re

```diff
@@ -1,84 +1,105 @@
+            "error": error =>
+              self.send(
+                UpdateResponse(
+                  Failed(Js.Exn.message(error) |> Js.Option.getWithDefault(""))
+                )
+              )
+          })
+        );
+      ();
     };
-
-    type action =
-      | Result(string)
-      | Error(string);
-
-    let sendQuery = (~query, ~reduce) => {
-      let _ =
-      Js.Promise.(
-        resolve(InternalConfig.apolloClient##query({
-          "query": [@bs] gql(query##query),
-          "variables": query##variables
-        }))
-        |> then_(
-             (value) => {
-               reduce(() => Result(value), ());
-               resolve()
-             }
-           )
-        |> catch(
-             (_value) => {
-               reduce(() => Error("an error happened"), ());
-               resolve()
-             }
-           )
-      );
+    let removeSubscription = self => {
+      switch self.ReasonReact.state.subscription^ {
+      | Some(subscription) => subscription##unsubscribe()
+      | None => ()
+      };
+      self.state.subscription := None;
     };
-
-    let component = ReasonReact.reducerComponent("ReasonApollo");
-    let make = (~query as q, children) => {
+    {
       ...component,
       initialState: () => {
-        response: Loading,
+        let queryObservable = makeQueryObservable(~query=q);
+        let response = convertResponse(queryObservable##currentResult());
+        {
+          response,
+          queryObservable: ref(queryObservable),
+          subscription: ref(None)
+        };
+      },
+      retainedProps: {
         variables: q##variables
       },
       reducer: (action, state) =>
         switch action {
-          | Result(result) => {
-            let typedResult = castResponse(result)##data;
-            ReasonReact.Update({
-              ...state,
-              response: Loaded(typedResult)
-            })
-          }
-          | Error(error) => ReasonReact.Update({
-            ...state,
-            response: Failed(error)
-          })
+        | UpdateResponse(response) => Update({...state, response})
         },
-      willReceiveProps: ({state, reduce}) => {
-        if(!shallowEqual(asJsObject(q##variables), asJsObject(state.variables))) {
-          sendQuery(~query=q, ~reduce);
-          state;
-        } else {
-          state;
-        }
-      },
-      didMount: ({reduce}) => {
-        sendQuery(~query=q, ~reduce);
+      didMount: self => {
+        startSubscription(self);
         ReasonReact.NoUpdate;
       },
-      render: ({state}) => {
-        children(state.response, q##parse);
-      }
+      willReceiveProps: self =>
+        if (! shallowEqual(q##variables, self.retainedProps.variables)) {
+          removeSubscription(self);
+          let queryObservable = makeQueryObservable(~query=q);
+          self.state.queryObservable := queryObservable;
+          startSubscription(self);
```

Comment:
I think the warning happens because this triggers a synchronous setState when data is in cache.